src/iotjs_module_httpparser.cpp

```diff
@@ -0,0 +1,483 @@
+/* Copyright 2015 Samsung Electronics Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "iotjs_def.h"
+#include "iotjs_module_httpparser.h"
+#include "http_parser.h"
+#include "iotjs_objectwrap.h"
+#include "iotjs_module_buffer.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+
+namespace iotjs {
+
+
+#define JSETPROPERTY( container_, fname_, vname_ )                      \
+  do {                                                                  \
+    JObject jobj(vname_ );                                              \
+    container_.SetProperty(fname_, jobj);                               \
+  } while(0)
+
+
+// max header size. If header field > HEADER_MAX, flush header to JS side.
+// increse this to minimize inter JS-C call
+#define HEADER_MAX 10
+
+
+
+class HTTPParserWrap : public JObjectWrap {
+public:
+  explicit HTTPParserWrap(JObject& parser_, http_parser_type type)
+    : JObjectWrap(parser_) {
+    Initialize(type);
+    parser.data = this;
+  }
+
+  ~HTTPParserWrap() {
+  }
+
+  void Initialize(http_parser_type type);
+
+  // http-parser callbacks
+  int OnMessageBegin() {
+    url.MakeEmpty();
+    status_msg.MakeEmpty();
+    return 0;
+  }
+
+  int OnUrl(const char* at, size_t length) {
+    url.Append(at, length);
+    return 0;
+  }
+
+  int OnStatus(const char* at, size_t length) {
+    status_msg.Append(at, length);
+    return 0;
+  }
+
+  int OnHeaderField(const char* at, size_t length) {
+
+    if (n_fields == n_values) {
+      n_fields++;
+      if (n_fields == HEADER_MAX) {
+        Flush(); // to JS world
+        n_fields = 1;
+        n_values = 0;
+      }
+      fields[n_fields-1].MakeEmpty();
+    }
+    IOTJS_ASSERT(n_fields == n_values + 1);
+    fields[n_fields-1].Append(at, length);
+
+    return 0;
+  }
+
+  int OnHeaderValue(const char* at, size_t length) {
+    if (n_fields != n_values) {
+      n_values++;
```

Comment:
```values``` and ```fields``` are flushed to JS in ```OnHeaderField``` function just before corresponding ```OnHeaderValue``` is called. So, it is not.